chain/rosetta-rpc/src/lib.rs

```diff
@@ -185,6 +185,80 @@ async fn network_options(
     }))
 }
 
+/// Get a block with `block_id`.
+/// Returns `Ok(Some(_))` if the block exists and is final.
+/// Returns `Ok(None)` if the block does not exist or is not final.
+async fn get_block_if_final(
+    block_id: &near_primitives::types::BlockReference,
+    view_client_addr: &Addr<ViewClientActor>,
+) -> Result<Option<near_primitives::views::BlockView>, models::Error> {
+    let mut is_query_by_height = false;
+    let mut is_query_final = false;
+    match block_id {
+        near_primitives::types::BlockReference::BlockId(
+            near_primitives::types::BlockId::Height(_),
+        ) => {
+            is_query_by_height = true;
+        }
+        near_primitives::types::BlockReference::Finality(
+            near_primitives::types::Finality::Final,
+        ) => {
+            is_query_final = true;
+        }
+        _ => {}
+    }
+    let block = match view_client_addr.send(near_client::GetBlock(block_id.clone())).await? {
+        Ok(block) => block,
+        Err(_) => return Ok(None),
+    };
+    if is_query_final {
+        return Ok(Some(block));
+    }
+    let final_block = match view_client_addr
+        .send(near_client::GetBlock(near_primitives::types::BlockReference::Finality(
+            near_primitives::types::Finality::Final,
+        )))
+        .await?
+    {
+        Ok(block) => block,
+        Err(_) => {
+            return Err(errors::ErrorKind::InternalError("final block not found".to_string()).into())
+        }
+    };
```

Comment:
We always make this call.  It might be more readable to restructure
this function to make this query first.  We can also put height check
here which gets rid of one fetch if querying non-final heights:

```
    let final_block = ...;
    let is_query_by_height = match block_id {
        ...::Finality(...) => return Some(final_block),
        ...::BlockId(...Height(height)) => {
            if height > final_block.header.height {
                return Ok(None);
            }
            true
        },
        _ => false
    };

    let block = ...;

    if is_query_be_height {
        return Ok(Some(block));
    }
    if block.header.height > final_block.header.height {
        return Ok(None);
    }

    ...
```